Issue #1535 - Add a `help` command (alias of `--help`)

timacias opened the pull request:
- [x] I have added a news fragment under `changelog.d/` (if the patch affects the end users)

## Summary of changes
1. `pipx --help` and `pipx [command_name] --help` both produce help info.
2. Issue #1535 requests a new pipx command that is an alias for `--help`.
3. `pipx help` and `pipx help [command_name]` can now be used to provide identical output to (1).
4. The help info now references the new help command.

## Test plan

<!-- provide evidence of testing, preferably with command(s) that can be copy+pasted by others -->

Tested by running

```
# command(s) to exercise these changes
python3 src/pipx help
python3 src/pipx help install
python3 src/pipx help pin

# if pipx is installed to your system
pipx help
pipx help install
pipx help pin
```

This is my first PR to a public repo. Please feel free to let me know if there are any guidelines or general tenets I've failed to follow. Thanks!

timacias committed:
Initial awareness of 'help' command

Initial addition of a `commands/help.py` file
Add "help" to `commands/__init__.py` to make pipx aware of a new command

timacias committed:
Alias help command to --help flag in main.py

Added an _add_help function which informs argparse of a "help" command, and an optional argument for it called "subcommand"
Added a check for args.command == "help"

timacias committed:
Added changelog for Issue #1535

timacias committed:
Add whitespace

timacias committed:
Remove unused import from __init__.py

timacias committed:
Add missing return statements to main.py

timacias committed:
Remove implicit string concat in main.py

timacias committed:
Update main.py

Reflects requirements set by the linter

timacias committed:
Merge branch 'pypa:main' into Open-source-Contribution

Gitznik commented on src/pipx/main.py:
```diff
@@ -973,7 +997,9 @@ def get_command_parser() -> Tuple[argparse.ArgumentParser, Dict[str, argparse.Ar
 
-    subparsers = parser.add_subparsers(dest="command", description="Get help for commands with pipx COMMAND --help")
+    subparsers = parser.add_subparsers(
+        dest="command", description="Get help for commands with pipx COMMAND --help\n\t\t\t  or pipx help COMMAND"
```
Please stick to how the other commands wrap text in their description.

Gitznik reviewed (changes requested):
Thanks for your contribution :muscle: can you please add some tests for this?
Also can you verify this will work with commands with multiple subcommands, like `pipx interpreter list`?

timacias committed:
Merge branch 'pypa:main' into Open-source-Contribution

timacias committed:
help subcommand now works with multiple arguments

Example: `pipx help interpreter list` now has an identical output to `pipx interpreter list --help`

timacias commented:
Thanks for your comment. I have added a fix for cases with multiple subcomamnds (e.g., `pipx interpreter list` and `pipx interpreter prune`) . I'll add tests for the help subcommand in my next commit.

timacias committed:
Merge branch 'pypa:main' into Open-source-Contribution

timacias committed:
Remove redundant else branch for help command

timacias committed:
Add unit tests for the help subcommand

The line `valid_commands = parser._subparsers._actions[4].choices.keys()` causes the linter to throw an error:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tests/test_help.py:23: error: Item "None" of "_ArgumentGroup | None" has no attribute "_actions"  [union-attr]
tests/test_help.py:23: error: Item "Iterable[Any]" of "Iterable[Any] | Any | None" has no attribute "keys"  [union-attr]
tests/test_help.py:23: error: Item "None" of "Iterable[Any] | Any | None" has no attribute "keys"  [union-attr]
Found 3 errors in 1 file (checked 76 source files)
```

All linter tests pass normally when that line is substituted with a hard-coded list of valid pipx commands.

totall4u commented:
> - [x] I have added a news fragment under `changelog.d/` (if the patch affects the end users)
> 
> 
> 
> ## Summary of changes
> 
> 1. `pipx --help` and `pipx [command_name] --help` both produce help info.
> 
> 2. Issue #1535 requests a new pipx command that is an alias for `--help`.
> 
> 3. `pipx help` and `pipx help [command_name]` can now be used to provide identical output to (1).
> 
> 4. The help info now references the new help command.
> 
> 
> 
> ## Test plan
> 
> 
> 
> <!-- provide evidence of testing, preferably with command(s) that can be copy+pasted by others -->
> 
> 
> 
> Tested by running
> 
> 
> 
> ```
> 
> # command(s) to exercise these changes
> 
> python3 src/pipx help
> 
> python3 src/pipx help install
> 
> python3 src/pipx help pin
> 
> 
> 
> # if pipx is installed to your system
> 
> pipx help
> 
> pipx help install
> 
> pipx help pin
> 
> ```
> 
> 
> 
> This is my first PR to a public repo. Please feel free to let me know if there are any guidelines or general tenets I've failed to follow. Thanks!

Gitznik reviewed (changes requested):
Generally, looks fine for me. Please make all pipelines pass, I'll give it another proper review once that's done.

timacias committed:
Merge branch 'main' into Open-source-Contribution

timacias committed:
Use an explicit list of valid pipx commands

Rather than accessing a protected attribute of `parser._subparsers`

Note: a commented out line with the protected access is still included in the event that the list of valid pipx commands changes

Gitznik reviewed (changes requested):
Just played around with this locally. If running `pipx help`, the autocomplete does not work anymore unfortunately. Could you have a look at that please?

timacias committed:
Merge branch 'pypa:main' into Open-source-Contribution

timacias committed:
Merge branch 'main' into Open-source-Contribution

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gaborbernat reviewed (changes requested):
Can you rebase and fix the merge conflicts?

gaborbernat commented:
Seems stalled we can re-open if you get back to it.
